### PR TITLE
Fix dual dof handler sparsity

### PR DESCRIPTION
**Pull request by drwells**
This is the fix for an issue mentioned on the mailing list where step-35 style sparsity pattern generation failed in parallel. That particular version of `make_sparsity_pattern` calls `GridTools::get_finest_common_cells` which doesn't work in parallel.

I am pretty sure that the current fix is right but I want to do a few more checks (and add another test to explicitly test the `GridTools` function), so this is still WIP for now.

### CODE REVIEW

**Comment by masterleinad** [comment]
At least, you need to fix the indentation. :wink:

**Comment by drwells** [comment]
I think this is now done: the generated sparsity pattern couples all dofs on active cells of dof handler 0 to all dofs on the same cells of dof handler 1.

Since, after grid refinement, some cells of `p::d::Triangulation` may be moved to another processor, I don't think its possible to implement the `GridTools` function in a reasonable way in the completely distributed setting: hence, I banned using two different `p::d::Triangulation`s. However, things should work for `p::s::Triangulation` so I added a test for it.

**Comment by drwells** [comment]
I managed to break the dof handler test when fixing the sparsity pattern test; I am fixing this now.

**Comment by drwells** [comment]
That should fix it.

**Comment by drwells** [comment]
The current fix isn't the right one with shared triangulations; I am looking at this now.

**Comment by drwells** [comment]
Okay, that should fix that issue.

**Comment by masterleinad** [comment]
```
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-7763/source/grid/grid_tools_dof_handlers.cc:1138:19: error: unused variable 'dim' [-Werror,-Wunused-variable]
    constexpr int dim      = MeshType::dimension;
                  ^
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-7763/source/grid/grid_tools_dof_handlers.cc:1139:19: error: unused variable 'spacedim' [-Werror,-Wunused-variable]
    constexpr int spacedim = MeshType::space_dimension;
                  ^
```

**Comment by drwells** [comment]
Fixed.

File: source/dofs/dof_tools_sparsity.cc [hunk 1]
```diff
@@ -222,23 +225,62 @@ namespace DoFTools
...
+    typename std::list<std::pair<cell_iterator, cell_iterator>>::const_iterator
+      cell_iter = cell_list.begin();
 
     for (; cell_iter != cell_list.end(); ++cell_iter)
       {
-        const typename DoFHandlerType::cell_iterator cell_row =
-          cell_iter->first;
-        const typename DoFHandlerType::cell_iterator cell_col =
-          cell_iter->second;
+        const cell_iterator cell_row = cell_iter->first;
```
Comment (masterleinad) [inline source/dofs/dof_tools_sparsity.cc]: Can't we use a range-based for loop here?

Comment (drwells) [reply source/dofs/dof_tools_sparsity.cc]: We can. I converted all of the typedefs but didn't convert the loop. I'll fix it.

**Pull request merged**